Add node size scaling to the Graphviz output for the high level graphs

freyam opened the pull request:
- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask` / `isort dask`

/cc @martindurant @GenevieveBuckley

Hi. I created this PR to add new size features to the HighLevelGraphs Graphviz depiction based on the amount of tasks each HLG layer has.

This is a follow-up to the information found at https://github.com/dask/dask/issues/7847/.

For now, I have just added a simple implementation of just a trivial case.

At this point in the code, adding new conditions and loops is simple and quick. All that remains is to decide how we want to display this information on the HLG.

GenevieveBuckley commented on dask/highlevelgraph.py:
```diff
@@ -1219,6 +1219,22 @@ def to_graphviz(
         attrs.setdefault("shape", "box")
+        n_tasks = len(hg.layers[k])
+        # print(n_tasks)
+        if n_tasks <= 10:
```
I would like to see fontsize as a continuous variable, rather than these discrete "buckets" for different ranges.

If it's a bit too tricky for you to do this with log scaling, a good first step might be to make an implementation with a linear relationship between ntasks and fontsize.

freyam replied:
Oh yes. I thought of the same. For now, I just wrote something very simple. 

Once our plan is final, I will make it clean.

freyam commented:
It looks like this right now:

![graph-high](https://user-images.githubusercontent.com/62539811/124760946-4eee2e80-df42-11eb-89f4-a36cbd04936a.png)

It looks unappealing to me. Maybe, we can remove the row and just add it within a bracket.

![image](https://user-images.githubusercontent.com/62539811/124766100-a8a52780-df47-11eb-8a2d-ee83645e4705.png)
The top one in this image.

martindurant commented:
I would certainly opt for `N=100` or `100tasks`, whether in a bracket or on a separate line.If on a separate line, maybe I would suggest the same font size for this part of the box for all boxes.

freyam commented:
@martindurant 
I prefer the bracket variant. 

looks clean and we will also be able to add more later in a different layer. The idea you suggested about having IO | CPU | GPU and so on can be shown on a different row.

martindurant commented:
I would suggest that "type" is the thing that makes most sense in colour, whether box background, border or font.

freyam commented:
Yes! I was planning to make the `dask.array` tasks in red (red border with a light red fill color) and `dask.dataframes` to have a green/blue color (green/blue border with light green/blue fill color).

and the standard Black and White color (and fill color) for the remaining tasks.

This would look really pleasing to the eyes as well. But, as @GenevieveBuckley suggested, I will wait this out till we get more textual information on the graph. We can add color at the end.

martindurant commented:
Agreed, but I was referring to your comment

> The idea you suggested about having IO | CPU | GPU and so on can be shown on a different row.

... I would not make too many rows. The basic type information is the class of the layer (there are a few of these) and the above three categories couple only loosely to API (array/dataframe/X). I don't think colouring by API is particularly interesting, there will only normally be one of these in the graph at a time, and it's often in the layer names anyway.

freyam commented:
>  I don't think colouring by API is particularly interesting, there will only normally be one of these in the graph at a time, and it's often in the layer names anyway.

Yes I agree :heavy_check_mark: 

What do you suggest for this particular PR?
Can I replace the `-#x` part from the `name-#x` which is currently followed to make it 
`name-#x` -> `name#n_tasks`
(remove the `-` and change the number to `n_tasks`)

freyam commented:
![image](https://user-images.githubusercontent.com/62539811/124788214-35f17780-df5a-11eb-9792-7d5f429de9c0.png)
this?

martindurant commented:
No, I would not change the name of the layer, but display that name in full. The character "#" means *ordinal* number, or (numerical) tag, not quantity.

Note that in your examples, you should really use log scaling to give us a better idea of how it will eventually look. May as well implement that in code right now and show some real examples.

freyam commented:
Oh yes, I have already added that implementation. I normalized the range of `n_tasks` : `fontsize`.
The examples I am showing are from the real source produced by `to_graphviz()`. I am just using http://magjac.com/graphviz-visual-editor/ to show it to you all.

You can have a look at the latest commit. I decided not to use `log` scaling and just used normalized values for all the tasks.

### The Math

Minimum size of the Node: 10 (This gets mapped to the layer with the least tasks)
Maximum size of the Node: 30 (This gets mapped to the layer with the most tasks)

```
fontsize = 10 + (n_tasks - minimum_n_tasks)/(maximum_n_tasks - minimum_n_tasks) * 20
```
`(n_tasks - minimum_n_tasks)/(maximum_n_tasks - minimum_n_tasks)` -> [0, 1]
So, `fontsize` -> [10, 30] 


I like this method and I think it locally normalizes all kinds of graph needs. Some graphs will have different minimums and different maximums and they will still look nice.

The examples I have sent till now all have been implemented.

freyam commented:
## Current Implementation

```py
from dask.datasets import timeseries

result = timeseries().shuffle("id", shuffle="tasks").head(compute=False)

# result.visualize(filename="graph-low")
result.dask.visualize(filename="graph-high")

```

![graph-high](https://user-images.githubusercontent.com/62539811/124792224-deeda180-df5d-11eb-96ac-c9db908ee1e8.png)


```py
import dask.array as da

a = da.ones((20), chunks=(2))

b = da.dot(a, a) + da.dot(a, a) + da.dot(a, a)
result = da.dot(a, a) + da.dot(a, a) + da.dot(a, a)

result = result + 10

result = result + da.dot(a, a)

result = result - 5

# result.visualize(filename="graph-low")
result.dask.visualize(filename="graph-high")
```

![graph-high](https://user-images.githubusercontent.com/62539811/124792387-05abd800-df5e-11eb-83a4-a2fe8f949e61.png)

```py
import dask.array as da

x = da.arange(1000).reshape(1, 1000)
y = x.T
z = x * y
result = da.sum(z)

# result.visualize(filename="graph-low")
result.dask.visualize(filename="graph-high")
```

![graph-high](https://user-images.githubusercontent.com/62539811/124792677-43a8fc00-df5e-11eb-8446-5a125b3f90de.png)

freyam commented:
The only thing that is left now is how we want to show the `n_tasks` number.
It's easy to change that. So, if anyone has any ideas, lemme know. I will add it quickly and send demos.

martindurant commented:
OK, the code seems to do the right thing (your example using fixed values was not using this calculation).

freyam commented:
Oh yes! my bad :) That was an old screenshot.

freyam commented:
/cc @mrocklin 
I'd love to get your thoughts on this.
How would you like to show the `n_tasks` on the final output?

mrocklin commented:
In general this looks good to me.  A couple of suggestions:

1. For sizing I like that we're operating within a fixed range and that we're responsive to min/max of the graph.  We might consider increasing the dynamic range of the response curve a bit from something linear to something with a log in it.  To be clear I'm not suggesting this, but rather only suggesting that we consider it. This can also be future work, since our objective is not well defined here yet 
2.  It's not clear to me that we want to print out the number of tasks.  We may want this but I think that we should plan this out a bit more before doing this.  I recommend that for now we only convey this information through the size, which is a bit more subtle.  We should plan to continue the conversation of how to render nodes in the issue though.

freyam commented:
> I recommend that for now we only convey this information through the size, which is a bit more subtle.

Whenever you used "size", I always assumed it meant number of tasks. It feels like I was wrong!

Do you mean the chunk size? partition size? dimension?

mrocklin commented:
Sorry, what I meant to say is "I recommend that we only convey the number
of tasks through the size of the node in graphviz.  I recommend that we not
currently also print out the number of tasks in the graphviz node."

Terms like chunk size, partition size, and dimension are specific to the
different collections (dask array, dask dataframe, ...) and aren't fully
general.  We probably can't use those concepts for the general visualize
method, but we can start using them when we being to specialize the
representation of each node based on the type of the layer.

On Wed, Jul 7, 2021 at 4:06 PM Freyam Mehta ***@***.***>
wrote:

> I recommend that for now we only convey this information through the size,
> which is a bit more subtle.
>
> Whenever you used "size", I always assumed it meant number of tasks. It
> feels like I was wrong!
>
> Do you mean the chunk size? partition size? dimension?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/7869#issuecomment-875932635>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACKZTAUFNRPLTLEFH2SZTDTWS6T5ANCNFSM472MWE2A>
> .
>

freyam commented:
Oh, understood. I misunderstood it before. I will remove the extra text from the output.

freyam commented:
## Current Implementation

![1](https://user-images.githubusercontent.com/62539811/124867598-7212ef80-dfcf-11eb-9366-66bef5b61afa.png)
![2](https://user-images.githubusercontent.com/62539811/124867600-72ab8600-dfcf-11eb-81b4-bc33386be595.png)
![3](https://user-images.githubusercontent.com/62539811/124867601-73441c80-dfcf-11eb-8069-58b445c97b9e.png)

GenevieveBuckley commented:
It certainly looks cleaner/prettier without the n_tasks text. 

Perhaps I should reframe my thinking so that:

* graphviz output: good for getting a general sense of things
* HTML repr: good for getting a more detailed overview

Separate to this PR, we should consider adding `n_tasks` to the `collection_annotations` dictionary, so that it can displayed in the HTML repr. We're not doing that right now, mostly because some layers (like the `ArrayOverlapLayer`) materialize when you call `len()`. That shouldn't be a blocker though.

GenevieveBuckley commented on dask/highlevelgraph.py:
```diff
@@ -1207,18 +1207,28 @@ def to_graphviz(
     for k in hg.dependencies:
         k_name = name(k)
         attrs = data_attributes.get(k, {})
+        xfontsize = int(10 + ((n_tasks_dict[k] - mn) / mx) * 20)
```
This is a much nicer implementation than the first one you had :smile:

freyam replied:
Yayyy :rocket:

freyam replied:
Yayy! :rocket:

freyam commented:
Is there something else that needs to be done here?

I fixed the formula a bit (took care of a corner case) and cleaned the code a little.

## Current Implementation (Increased minimum size to make it more readable)

![1](https://user-images.githubusercontent.com/62539811/125156143-08cce100-e175-11eb-9c44-48266576efc7.png)
![2](https://user-images.githubusercontent.com/62539811/125156145-09fe0e00-e175-11eb-9823-ab52a1fa098b.png)
![3](https://user-images.githubusercontent.com/62539811/125156146-09fe0e00-e175-11eb-956c-a2b3cb159a8e.png)
![4](https://user-images.githubusercontent.com/62539811/125156147-0a96a480-e175-11eb-97e3-45f2955d850c.png)

freyam commented:
## Example 4 from the previous comment

This example really glorifies the changes I have made. 

### Code

```py
import dask.array as da

x = da.random.random((10000, 10000), chunks=(200, 200))
result = x + x.T - x.mean(axis=0)

result.dask.visualize()
```

|        task_name        | n_tasks | fontsize |
|:-----------------------:|:-------:|:--------:|
|          sub-#0         |   2500  |    40    |
|          add-#1         |   2500  |    40    |
|     random_sample-#2    |   2500  |    40    |
|       transpose-#3      |   2500  |    40    |
|      mean_chunk-#4      |   2500  |    40    |
| mean_combine-partial-#5 |   650   |    24    |
| mean_combine-partial-#6 |   200   |    21    |
|  mean_agg-aggregate-#7  |    50   |    20    |


### Before (Current)
![image](https://user-images.githubusercontent.com/62539811/125156315-e7b8c000-e175-11eb-9e7b-856539ae7a90.png)

### After (My changes)

![image](https://user-images.githubusercontent.com/62539811/125156334-fdc68080-e175-11eb-9f68-288663e5afaf.png)


---------

Just for fun
What if we try changing the font to a Sans Serif instead of Serif one. 
When nodes have a fontsize of 40, they tend to look a little unpleasant.

++ HTML Repr also uses a Sans font.
![4](https://user-images.githubusercontent.com/62539811/125156459-952bd380-e176-11eb-98f4-6ff1870234aa.png)

This can be taken up in another PR. We can discuss the font family. Just sent this as this is something I noticed.

GenevieveBuckley commented:
> Is there something else that needs to be done here?

Not at the moment, no. Feel free to transfer your focus onto the next thing.

The one failed CI test doesn't look like it has anything to do with this PR. I think it's just flaky & will come good if the CI is re-triggered.

jrbourbeau commented:
Just re-triggered CI

GenevieveBuckley commented:
Well, I like it. Now that the CI tests have passed, can someone with merge privileges take a final look at it (maybe @martindurant  or @jrbourbeau) 

I'm mildly supportive of the suggestion to use a sans serif font for consistency with other representations of the high level graph, but I wouldn't die on this hill if other people dislike it. Now that the font size is often a lot bigger, it certainly seems more like a bunch of headings (where sans serif is typically expected) rather than paragraph-like text.

Re: this comment
> For sizing I like that we're operating within a fixed range and that we're responsive to min/max of the graph. We might consider increasing the dynamic range of the response curve a bit from something linear to something with a log in it. To be clear I'm not suggesting this, but rather only suggesting that we consider it. This can also be future work, since our objective is not well defined here yet

Now that we're not going to display the actual number of tasks in the text, I think it's worth sticking with a linear scaling until it's clear there is a problem a log or non-linear scale could fix. We can keep this idea in our back pocket until then. 

Linear scales are more intuitive for people so if we don't have to introduce another complicating factor, then I think we should keep things simple.

GenevieveBuckley commented:
@freyam can you change the PR title to something a bit clearer? Perhaps something like "Add size scaling to graphviz output for high level graphs"

(You probably remember we talked about how the PR title becomes the commit message since Dask squashes when merging PRs, so it's good to make it as clear as possible)

freyam commented:
Done :heavy_check_mark:

freyam commented:
> I'm mildly supportive of the suggestion to use a sans serif font for consistency with other representations of the high level graph, but I wouldn't die on this hill if other people dislike it. Now that the font size is often a lot bigger, it certainly seems more like a bunch of headings (where sans serif is typically expected) rather than paragraph-like text.

:raised_hands:

GenevieveBuckley commented:
bump @dask/maintenance

freyam commented:
💛💜

freyam commented:
@GenevieveBuckley Should I open a new PR for discussing the font family we would be using for Graphviz?
